### PR TITLE
implicitly restarted lanczos method for jax backend

### PR DESCRIPTION
**Pull request by mganahl**
depends on #834 and #832

### CODE REVIEW

File: tensornetwork/backends/jax/jax_backend_test.py [hunk 1]
```diff
@@ -669,6 +675,26 @@ def mv(x, H):
...
+  np.random.seed(10)
+  init = backend.randn((D,), dtype=dtype, seed=10)
+  H = backend.randn((D, D), dtype=dtype, seed=10)
+
+  def mv(x, H):
+    return jax.numpy.dot(H, x)
+
+  eta, U = backend.eigs(
+      mv, [H], init, numeig=4, num_krylov_vecs=40, maxiter=500, which=which,
+      tol=1E-5)
```
Comment (alewis) [inline tensornetwork/backends/jax/jax_backend_test.py:692]: Is it ok that the tolerance is fixed independently of the dtype?

File: tensornetwork/backends/jax/jitted_functions.py [hunk 2]
```diff
@@ -1,11 +1,42 @@
...
+      krylov_vectors: Matrix of krylov vectors, each row is treated as a
+        vector.
+      iterations: Number of iterations.
+    Returns:
+      jax.ShapedArray: The orthogonalized vector.
+    """
+    vec = vector
+    overlaps = 0
+    for _ in range(iterations):
+      ov = jax.numpy.dot(
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:31]: isn't this what vdot does?

**Review by alewis — COMMENTED** [review]
It's quite a lot of code so I may have missed something, but it looks good to me overall.

File: tensornetwork/backends/jax/jax_backend_test.py (hunk 1 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jax_backend_test.py:692]: in this case they are both double precision, so its fine. For lower precision one needs to change it, but this is currently not tested.

File: tensornetwork/backends/jax/jitted_functions.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py]: yes, thanks for catching that!

**Comment by mganahl** [comment]
> It's quite a lot of code so I may have missed something, but it looks good to me overall.

Once the other PRs are in, it'll be less

File: tensornetwork/backends/jax/jitted_functions.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:33]: actually, vdot only does scalar products

File: tensornetwork/backends/jax/jitted_functions.py [hunk 3]
```diff
@@ -185,19 +201,133 @@ def body_vector(i, vals):
...
+          (i >= jax.numpy.arange(Vm.shape[0]))[:, None] * Vm, precision)
+      alphas = alphas.at[i].set(overlaps[i])
+      norm = jax.numpy.linalg.norm(Av)
+      Av = jax.numpy.reshape(Av, shape)
+      # only normalize if norm is larger than threshold,
+      # otherwise return zero vector
+      Av = jax.lax.cond(norm > tol, lambda x: Av/norm, lambda x: Av * 0.0, None)
+      Vm, betas = jax.lax.cond(
+          i < num_krylov_vecs - 1,
+          lambda x: (Vm.at[i + 1, :].set(Av.ravel()), betas.at[i].set(norm)), #pylint: disable=line-too-long
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:309]: I think you can remove this disable.

**Review by Thenerdstation — APPROVED** [review]

**Review by alewis — APPROVED** [review]

**Pull request merged**